crypto/ec/ec_asn1.c

```diff
@@ -792,18 +794,79 @@ EC_GROUP *EC_GROUP_new_from_ecparameters(const ECPARAMETERS *params)
         goto err;
     }
 
+    /*
+     * Check if the explicit parameters group just created matches one of the
+     * built-in curves.
+     *
+     * We create a copy of the group just built, so that we can remove optional
+     * fields for the lookup: we do this to avoid that one of the optional
+     * parameters is used to force the library into using a less performant and
+     * less secure EC_METHOD instead of the specialized one.
+     * In any case, `seed` is not really used in any computation, while a
+     * cofactor different from the one in the built-in table is just
+     * mathematically wrong anyway and should not be used.
+     */
+    if (NULL == (ctx = BN_CTX_new())) {
+        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, ERR_R_BN_LIB);
+        goto err;
+    }
+    if (NULL == (dup = EC_GROUP_dup(ret))
+            || 1 != EC_GROUP_set_seed(dup, NULL, 0)
+            || !EC_GROUP_set_generator(dup, point, a, NULL)) {
+        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, ERR_R_EC_LIB);
+        goto err;
+    }
+    if (NID_undef != (curve_name = ec_curve_nid_from_params(dup, ctx))) {
+        /*
+         * The input explicit parameters successfully matched one of the
+         * built-in curves: often for built-in curves we have specialized
+         * methods with better performance and hardening.
+         *
+         * In this case we replace the `EC_GROUP` created through explicit
+         * parameters with one created from a named group.
+         */
+        EC_GROUP *named_group = NULL;
+
+#ifndef OPENSSL_NO_EC_NISTP_64_GCC_128
+        /*
+         * NID_wap_wsg_idm_ecid_wtls12 and NID_secp224r1 are both aliases for
+         * the same curve, we prefer the SECP nid when matching explicit
+         * parameters as that is associated with a specialized EC_METHOD.
+         */
+        if (curve_name == NID_wap_wsg_idm_ecid_wtls12)
+            curve_name = NID_secp224r1;
```

Comment:
Does this mean that if the group gets written out again it will change name (different OID)? Or is the OID still the same?

Reply:
We set the EXPLICIT_PARAMS ASN1 flag, so when it is serialized by default it comes out with the explicit parameters and no OID.

If one forces to output the OID, yes it will come out with the P-224 OID (but before this change it just gave an error, as we reach this only from an ECPARAMETERS object or parsing an explicit parameters file, so no OID was associated with the group at all and the library barked)

Reply:
Ah! Of course! I forgot. Explicit params obviously means no oid!!

Reply:
I wonder if this logic would be better in ec_curve_nid_from_params?

Reply:
I was unsure about that: `ec_curve_nid_from_params` is used also in the FIPS related checks, and there might be reason to avoid enforcing anything by default.

Truth is that currently this branch is a noop anyway because of the sorting of the built-in table: the `wap*` aliases come later in the table, so we already match the P-224 NID first anyway.

Reply:
@mattcaswell should I move this anyway into `ec_curve_nid_from_params`?

(Pinging just because Github UI tends to somehow hide the above comment for me)

Reply:
Ok - fine. Leave as is.